models/intermediate/int_klaviyo__event_attribution.sql

```diff
@@ -100,18 +100,73 @@ attribute as (
     from last_touches 
 ),
 
-final as (
-
+order_seq as (
     select
-        *,
-
-        -- get whether the event is attributed to a flow or campaign
-        coalesce(touch_type, first_value(touch_type) over(
-            partition by person_id, source_relation, touch_session order by occurred_at asc rows between unbounded preceding and current row)) 
+        attribute.*,
+        -- stable ordering in case of same-time events
+        sum(case when lower(attribute.type) = 'placed order' then 1 else 0 end) over (
+            partition by attribute.person_id, attribute.source_relation
+            order by attribute.occurred_at, attribute.unique_event_id
+            rows between unbounded preceding and current row
+        ) as placed_order_group
+    from attribute
+),
 
-            as session_touch_type -- if the session events qualified for attribution, extract the type of touch they are attributed to
+placed_order_index as (
+    -- one row per Placed Order carrying the touch to copy
+    select
+        order_seq.person_id,
+        order_seq.source_relation,
+        order_seq.placed_order_group,
+        order_seq.occurred_at as placed_order_occurred_at,
+        order_seq.last_touch_id as placed_order_touch_id
+    from order_seq
+    where lower(order_seq.type) = 'placed order'
+    ),
+
+adjust_orders as (
+    select
+        order_seq.*,
+        case
+            when
+                -- list of possible Shopify order events --> https://help.klaviyo.com/hc/en-us/articles/115005080447
+                lower(order_seq.type) in ('fulfilled order', 'fulfilled partial order', 'delivered shipment', 'marked out for delivery', 'confirmed shipment', 'cancelled order', 'refunded order') 
+                and order_seq.placed_order_group > 0
+                and {{ dbt.datediff('placed_order_index.placed_order_occurred_at', 'order_seq.occurred_at', 'month') }} <= 3
```

Comment:
I am 99.9% sure that if `placed_order_index.placed_order_occurred_at` is null this will work OK and the clause will be treated basically as false, but perhaps we wanna add a coalesce like this to be safe?
```sql
coalesce({{ dbt.datediff('placed_order_index.placed_order_occurred_at', 'order_seq.occurred_at', 'month') }}, 100) <= 3
```
I'll leave that up to you 🤓

Reply:
Ahh true--better to be safe. Added!